core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -309,24 +309,23 @@ class ZooKeeperClient(connectString: String,
   def sessionId: Long = inReadLock(initializationLock) {
     zooKeeper.getSessionId
   }
-
+  
   private def initialize(): Unit = {
-    if (!zooKeeper.getState.isAlive) {
+    if (!connectionState.isAlive) {
+      zooKeeper.close()
       info(s"Initializing a new session to $connectString.")
       // retry forever until ZooKeeper can be instantiated
-      while (true) {
+      var connected = false
+      while (!connected) {
         try {
-          zooKeeper.close()
```

Comment:
I moved the `close` out of the loop since we would be calling it repeatedly on the same instance. One side-effect is that `InterruptedException` is now propagated instead of caught. This is consistent with what happens if `InterruptedException` is thrown from `Thread.sleep`. Having said that, what is the desired behaviour when `InterruptedException` is thrown? cc @junrao